### PR TITLE
Adapt to new Poloniex API.

### LINKED ISSUE #20: Poloniex legacy API has been discontinued.
The legacy API (https://poloniex.com/public) seems to have been turned off, now the Pricenode is getting error response codes and no market data.  Seems Poloniex [announced](https://support.poloniex.com/hc/en-us/articles/10745573683735-Legacy-API-Decommission) last year, perhaps just now they have decommissioned legacy support.

You can check it by trying the [legacy API](https://docs.legacy.poloniex.com/#public-http-api-methods): `curl "https://poloniex.com/public?command=returnTicker"`

Knowm XChange library still [uses](https://github.com/knowm/XChange/blob/937607eaee8909485475c29e73b6afe8cecb98ba/xchange-poloniex/src/main/java/org/knowm/xchange/poloniex/PoloniexExchange.java#L36) the [legacy endpoint](https://github.com/knowm/XChange/blob/937607eaee8909485475c29e73b6afe8cecb98ba/xchange-poloniex/src/main/java/org/knowm/xchange/poloniex/Poloniex.java#L18-L20).  There was [some discussion of it a year ago](https://github.com/knowm/XChange/issues/4541), but the new API has not yet been adopted in the code.

### Options:

1. What
2. Update the XChange code (learning curve..)
3. Wait for someone else to update the XChange code.
4. Discontinue use of Poloniex.
5. Write own adapter for Poloniex.

### PR DESCRIPTION
**Pull request by ghost**
Fixes #20

As the official Poloniex library is [broken](https://github.com/bisq-network/bisq-pricenode/issues/20), here is a custom Poloniex feed, inspired by prior work done by @rodvar in #15.

It provides the same functionality as the old Poloniex feed, while adapting to the new API.

(Plus, there were many unused dependencies in libs.versions.toml, I cleaned it up).

Sample run:

    Aug-22 17:43:17.769 [Timer-21] INFO  b.p.s.p.Poloniex: refresh took 490 ms. 
    Aug-22 17:43:17.769 [Timer-21] INFO  b.p.s.p.Poloniex: BTC/ETC: 5.94E-4 
    Aug-22 17:43:17.769 [Timer-21] INFO  b.p.s.p.Poloniex: BTC/DOGE: 2.427E-6 
    Aug-22 17:43:17.769 [Timer-21] INFO  b.p.s.p.Poloniex: BTC/ZEC: 9.76E-4 
    Aug-22 17:43:17.769 [Timer-21] INFO  b.p.s.p.Poloniex: BTC/DCR: 4.92E-4 
    Aug-22 17:43:17.770 [Timer-21] INFO  b.p.s.p.Poloniex: BTC/LTC: 0.002483 
    Aug-22 17:43:17.770 [Timer-21] INFO  b.p.s.p.Poloniex: BTC/XMR: 0.005596 
    Aug-22 17:43:17.770 [Timer-21] INFO  b.p.s.p.Poloniex: BTC/ETH: 0.06275 
    Aug-22 17:43:17.770 [Timer-21] INFO  b.p.s.p.Poloniex: BTC/DASH: 0.001013

### CODE REVIEW

File: src/main/java/bisq/price/spot/providers/Poloniex.java [hunk 1]
```diff
@@ -20,26 +20,76 @@
...
+        Set<ExchangeRate> exchangeRates = new HashSet<>();
+        ResponseEntity<PoloniexTicker[]> response =
+                new RestTemplate().getForEntity(
+                        UriComponentsBuilder
+                                .fromUriString(POLONIEX_URL).build()
+                                .toUri(),
+                        PoloniexTicker[].class);
+        PoloniexTicker[] tickers = response.getBody();
+        if (tickers != null) {
+            for (PoloniexTicker ticker : tickers) {
```
Comment (rodvar) [inline src/main/java/bisq/price/spot/providers/Poloniex.java]: consider replacing for and ifs with Arrays.stream(tickers).allMatch( ... )

**Review by rodvar — COMMENTED** [review]
I've also tested it. All tests pass now (its broken in main since Polinex test is not passing as the PR properly explains). 

Also I can see it getting poloniex rates back again now! :) 

       Aug.-23 16:06:41.935 [Timer-20] INFO  b.p.s.p.Poloniex: refresh took 480 ms. 
Aug.-23 16:06:41.935 [Timer-20] INFO  b.p.s.p.Poloniex: BTC/XMR: 0.005568 
Aug.-23 16:06:41.935 [Timer-20] INFO  b.p.s.p.Poloniex: BTC/ETH: 0.0629 
Aug.-23 16:06:41.935 [Timer-20] INFO  b.p.s.p.Poloniex: BTC/ZEC: 9.68E-4 
Aug.-23 16:06:41.935 [Timer-20] INFO  b.p.s.p.Poloniex: BTC/DASH: 0.001013 
Aug.-23 16:06:41.935 [Timer-20] INFO  b.p.s.p.Poloniex: BTC/DCR: 4.92E-4 
Aug.-23 16:06:41.935 [Timer-20] INFO  b.p.s.p.Poloniex: BTC/DOGE: 2.396E-6 
Aug.-23 16:06:41.935 [Timer-20] INFO  b.p.s.p.Poloniex: BTC/LTC: 0.002492 
Aug.-23 16:06:41.935 [Timer-20] INFO  b.p.s.p.Poloniex: BTC/ETC: 6.0E-4

File: src/main/java/bisq/price/spot/providers/Poloniex.java (hunk 1 above)
Comment (ghost) [reply src/main/java/bisq/price/spot/providers/Poloniex.java]: Thanks, code replaced by:

    Arrays.stream(Objects.requireNonNull(tickers))
        .filter(ticker -> ticker.isMatch(requestedCurrencies))
        .forEach(ticker -> exchangeRates.add(
                new ExchangeRate(ticker.getCurrency(), ticker.getPrice(), new Date(), this.getName())));

**Comment by ghost** [comment]
@alvasw I have updated this PR to migrate to WebClient, based on #26

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -20,26 +20,18 @@ spring-boot-starter-web-lib = { strictly = '2.5.6' }
...
 junit-jupiter = { module = 'org.junit.jupiter:junit-jupiter', version.ref = 'junit-jupiter-lib' }
 knowm-xchange-binance = { module = 'org.knowm.xchange:xchange-binance', version.ref = 'knowm-xchange-lib' }
-knowm-xchange-bitbay = { module = 'org.knowm.xchange:xchange-bitbay', version.ref = 'knowm-xchange-lib' }
 known-xchange-bitfinex = { module = 'org.knowm.xchange:xchange-bitfinex', version.ref = 'knowm-xchange-lib' }
 known-xchange-bitflyer = { module = 'org.knowm.xchange:xchange-bitflyer', version.ref = 'knowm-xchange-lib' }
 known-xchange-bitstamp = { module = 'org.knowm.xchange:xchange-bitstamp', version.ref = 'knowm-xchange-lib' }
 known-xchange-btcmarkets = { module = 'org.knowm.xchange:xchange-btcmarkets', version.ref = 'knowm-xchange-lib' }
-known-xchange-cexio = { module = 'org.knowm.xchange:xchange-cexio', version.ref = 'knowm-xchange-lib' }
 known-xchange-coinbasepro = { module = 'org.knowm.xchange:xchange-coinbasepro', version.ref = 'knowm-xchange-lib' }
 known-xchange-coinmarketcap = { module = 'org.knowm.xchange:xchange-coinmarketcap', version.ref = 'knowm-xchange-lib' }
```
Comment (alvasw) [inline gradle/libs.versions.toml]: known-xchange-coinmarketcap is unused too.

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -51,12 +43,11 @@ spring-boot-starter-webflux = { module = 'org.springframework.boot:spring-boot-s
...
 knowm-xchange-libs = [
-    'knowm-xchange-binance', 'knowm-xchange-bitbay', 'known-xchange-bitfinex',
-    'known-xchange-bitflyer', 'known-xchange-bitstamp', 'known-xchange-btcmarkets', 'known-xchange-cexio',
-    'known-xchange-coinbasepro', 'known-xchange-coinmarketcap', 'known-xchange-coinmate', 'known-xchange-coinone',
-    'known-xchange-exmo', 'known-xchange-hitbtc', 'known-xchange-huobi', 'known-xchange-independentreserve',
-    'known-xchange-kraken', 'known-xchange-luno', 'known-xchange-mercadobitcoin', 'known-xchange-paribu',
-    'known-xchange-poloniex', 'known-xchange-quoine'
+    'knowm-xchange-binance', 'known-xchange-bitfinex',
+    'known-xchange-bitflyer', 'known-xchange-bitstamp', 'known-xchange-btcmarkets',
+    'known-xchange-coinbasepro', 'known-xchange-coinmarketcap', 'known-xchange-coinone',
```
Comment (alvasw) [inline gradle/libs.versions.toml]: known-xchange-coinmarketcap is unused too.

File: src/main/java/bisq/price/spot/providers/Poloniex.java [hunk 4]
```diff
@@ -20,26 +20,67 @@
...
-import java.util.Set;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
 
 @Component
 class Poloniex extends ExchangeRateProvider {
-
+    // Supported fiat: -
+    // Supported alts: DASH, DCR, DOGE, ETC, ETH, LTC, XMR, ZEC
+    private static final String CURRENCIES = "DASH,DCR,DOGE,ETC,ETH,LTC,XMR,ZEC";
```
Comment (alvasw) [inline src/main/java/bisq/price/spot/providers/Poloniex.java]: You can convert this to a `List<String>` and iterate over it later.
```java
private static final List<String> SUPPORTED_CURRENCIES =
            List.of("DASH", "DCR", "DOGE", "ETC", "ETH", "LTC", "XMR", "ZEC");
```

File: src/main/java/bisq/price/spot/providers/Poloniex.java [hunk 5]
```diff
@@ -20,26 +20,67 @@
...
-        // Supported alts: DASH, DCR, DOGE, ETC, ETH, LTC, XMR, ZEC
-        return doGet(PoloniexExchange.class);
+        Set<String> requestedCurrencies = new HashSet<>(Arrays.asList(CURRENCIES.split(",")));
+        Set<ExchangeRate> exchangeRates = new HashSet<>();
+        PoloniexTicker[] tickers =
+                WebClient.create().get()
+                    .uri(POLONIEX_URL)
+                    .accept(MediaType.APPLICATION_JSON)
+                    .retrieve()
+                    .bodyToMono(PoloniexTicker[].class)
```
Comment (alvasw) [inline src/main/java/bisq/price/spot/providers/Poloniex.java]: `Mono`s are for streams returning one or no objects. `Flux`s are for multiple entries. You can change this to:
```java
Flux<PoloniexTicker> poloniexTickerFlux = WebClient.create()
                .get()
                .uri(POLONIEX_URL)
                .accept(MediaType.APPLICATION_JSON)
                .retrieve()
                .bodyToFlux(PoloniexTicker.class);
```

File: src/main/java/bisq/price/spot/providers/Poloniex.java [hunk 6]
```diff
@@ -20,26 +20,67 @@
...
+                .block(Duration.of(30, ChronoUnit.SECONDS));
+        Arrays.stream(Objects.requireNonNull(tickers))
```
Comment (alvasw) [inline src/main/java/bisq/price/spot/providers/Poloniex.java]: Afterward, you can do all the operations on the `Flux` and collect it into a `Set<ExchangeRates>`.
```java
        return poloniexTickerFlux.filter(PoloniexTicker::isSupportedCurrency)
                .map(PoloniexTicker::toExchangeRate)
                .collect(Collectors.toSet())
                .block();
```

File: src/main/java/bisq/price/spot/providers/Poloniex.java [hunk 7]
```diff
@@ -20,26 +20,67 @@
...
+        public String getCurrency() {
+            // symbol might not be a BTC pair, in which case return empty string
+            int splitIndex = symbol.indexOf("_BTC");
+            if (splitIndex < 0) {
+                return "";
+            }
+            return symbol.substring(0, splitIndex);
+        }
+
+        public boolean isMatch(Set<String> requestedCurrencies) {
```
Comment (alvasw) [inline src/main/java/bisq/price/spot/providers/Poloniex.java]: And simplify this for the `Flux` filter:
```java
        private boolean isBtcPair() {
            return symbol.endsWith("_BTC");
        }

        public boolean isSupportedCurrency() {
            return isBtcPair() && SUPPORTED_CURRENCIES.contains(getCurrency());
        }
```

File: src/main/java/bisq/price/spot/providers/Poloniex.java [hunk 8]
```diff
@@ -20,26 +20,67 @@
...
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    private static class PoloniexTicker {
+        public String symbol;
+        public String price;
+
+        public BigDecimal getPrice() {
+            return new BigDecimal(price);
+        }
+
+        public String getCurrency() {
```
Comment (alvasw) [inline src/main/java/bisq/price/spot/providers/Poloniex.java:74]: You can simplify this with `String.split(...)`.
```java
        public String getCurrency() {
            // DASH_BTC, DOGE_BTC, LTC_BTC, ...
            return symbol.split("_")[0];
        }
```

File: src/main/java/bisq/price/spot/providers/Poloniex.java [hunk 9]
```diff
@@ -20,26 +20,67 @@
...
+                .forEach(ticker -> exchangeRates.add(
+                        new ExchangeRate(ticker.getCurrency(), ticker.getPrice(), new Date(), this.getName())));
+        return exchangeRates;
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    private static class PoloniexTicker {
+        public String symbol;
+        public String price;
+
```
Comment (alvasw) [inline src/main/java/bisq/price/spot/providers/Poloniex.java:65]: You can add a `toExchangeRate()` for the flux filter above.
```java
        public ExchangeRate toExchangeRate() {
            return new ExchangeRate(getCurrency(), getPrice(), new Date(), "Poloniex");
        }
```

**Review by alvasw — CHANGES_REQUESTED** [review]
Good work!!

Nit

1. Do you know why we removed Coinbase in the past [[1](https://github.com/bisq-network/bisq-pricenode/commit/453931b114da6042d919c2f8e0524210b5610070)]?

2. `WebClient` supports JSON Arrays and can return these as `Flux` objects (see code review suggestions).

I pushed the suggestions too: https://github.com/alvasw/bisq-pricenode/commit/c958d250ded1a7ee556a7f39ce8e6b1d74ce60fd


[1] https://github.com/bisq-network/bisq-pricenode/commit/453931b114da6042d919c2f8e0524210b5610070

**Comment by ghost** [comment]
@alvasw, thanks for the detailed review!  Changes applied.

> 1. Do you know why we removed Coinbase  in the past [[1](https://github.com/bisq-network/bisq-pricenode/commit/453931b114da6042d919c2f8e0524210b5610070)]?

Perhaps Coinmarketcap was removed because the API switched to requiring a key.  As far as I'm aware we don't use any APIs that require keys.  [Disclaimer: don't know the actual history].

**Review by alvasw — APPROVED** [review]
utACK

**Review by gabernard — APPROVED** [review]
utACK

**Pull request merged**